src/Microsoft.DotNet.Build.Tasks/NuGetPush.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    /// <summary>
+    /// Run NuGet.exe to push a set of packages.
+    /// 
+    /// TimeoutPerPushInSeconds is a timeout per package push. This allows the task to retry when an
+    /// HTTP hang happens. A "NuGet.exe push *.nupkg -Timeout ###" call would fail without retrying.
+    /// https://github.com/NuGet/Home/issues/5224 requests retrying on hangs in NuGet.exe.
+    /// </summary>
+    public class NuGetPush : Task, ICancelableTask
+    {
+        [Required]
+        public string NuGetExePath { get; set; }
+
+        [Required]
+        public ITaskItem[] Inputs { get; set; }
+
+        [Required]
+        public string Source { get; set; }
+
+        public string ApiKey { get; set; }
+
+        public string Verbosity { get; set; } = "Detailed";
+
+        public int TimeoutPerPushInSeconds { get; set; } = TimeSpan.FromMinutes(5).Seconds;
+
+        public int RetriesPerPush { get; set; } = 5;
+
+        private bool _cancelRemainingPushes;
+
+        public void Cancel()
+        {
+            _cancelRemainingPushes = true;
+        }
+
+        public override bool Execute()
+        {
+            foreach (ITaskItem input in Inputs)
```

Comment:
To retry only the hanged package, we'd have to parse the logs to figure it out, then. The timeout logic would also be more complicated and potentially less reliable, since (unless we do even more real-time log parsing) we can only have one timeout per push call. I also haven't seen significant startup time issues, but I'll be getting some more info there.

I could try doing pushes in parallel--I've been told this won't work in the past, but I don't know what would go wrong. :smile: